### PR TITLE
Graphing and Reporting.

### PR DESCRIPTION
**Pull request by richardssam**
This is an update that includes some reporting code, as well as code to procedurally generate a test file for documentation.

### CODE REVIEW

File: enctests/testframework/generatetests.py [hunk 1]
```diff
@@ -0,0 +1,149 @@
...
+        help='Specify output file to write test wrappers to.'
+    )
+
+    args = parser.parse_args()
+
+    return args
+
+def main():
+    args = parse_args()
+    rootdir = os.path.abspath(args.root)
```
Comment (apetrynet) [inline enctests/testframework/generatetests.py:38]: We've used "pathlib" for all path handling in other parts of the code base. Should we be consistent and use it here as well?

File: enctests/testframework/generatetests.py [hunk 2]
```diff
@@ -0,0 +1,149 @@
...
+                    info = ""
+                else:
+                    incommand = True
+                    lastcommandstart = linenumber
+                continue
+            if incomment:
+                info = info + line
+            if incommand:
+                command = command + line
+
```
Comment (apetrynet) [inline enctests/testframework/generatetests.py:87]: We'll need to close the file here if we don't use the context manager approach.
`f.close()`

File: enctests/testframework/generatetests.py [hunk 3]
```diff
@@ -0,0 +1,149 @@
...
+        print("------\n", markdownfile)
+        f = open(markdownfile, "r")
+        incomment = False
+        incommand = False
+        command = ""
+        info = ""
+        lastcommandstart = -1
+        linenumber = 0
+        for line in f:
+            linenumber = linenumber + 1
```
Comment (apetrynet) [inline enctests/testframework/generatetests.py:58]: We can combine this:
```
for linenumber, line in enumerate(f):
```

File: enctests/testframework/generatetests.py [hunk 4]
```diff
@@ -0,0 +1,149 @@
...
+            testconfigs[testname]['wedges'] = {'wedge0': wedge}
+        tests.append(testconfigs)
+
+    reports = {'reports': {'name': 'doc-tests',
+                            'directory': 'docs-encode',
+                            'templatefile': 'doctests.html.jinja'}}
+    tests.append(reports)
+
+    f = open(args.output_config_file, "w")
+    yaml.dump_all(tests, f)
```
Comment (apetrynet) [inline enctests/testframework/generatetests.py:145]: Same comment regarding closing files. The recommended way to open/close files is:

```
with open(args.output_config_file, "w") as f:
    yaml.dump_all(tests, f)
```

File: enctests/testframework/main.py [hunk 5]
```diff
@@ -397,6 +404,10 @@ def main():
     # Serialize to *.otio
     otio.adapters.write_to_file(timeline, args.output)
 
+    # Generate any reports (if specified in file)
```
Comment (apetrynet) [inline enctests/testframework/main.py:407]: Should this be triggered/toggled by an argument? `--generate-html` for instance?

File: enctests/testframework/utils/outputTemplate.py [hunk 6]
```diff
@@ -0,0 +1,102 @@
...
+          # for key, enc_data in enc_info.items():
+          #     if key == 'results':
+          #         for arg, value in enc_data.items():
+          #             merge_test_info[arg] = value
+          #     else:
+          #         #print("Something ese:", value)
+          #         merge_test_info['encoder_version'] = key
+          #         for arg, value in enc_data.items():
+          #             merge_test_info[arg] = value
+          #print("testInfo:", test_info.metadata)
```
Comment (apetrynet) [inline enctests/testframework/utils/outputTemplate.py:51]: We should perhaps remove commented blocks like this and rather bring it back to life with git if we regret.

File: enctests/testframework/utils/outputTemplate.py [hunk 7]
```diff
@@ -0,0 +1,102 @@
...
+
+      filename = reportconfig['name']+"-"+graph.get("name")
+      if "directory" in reportconfig:
+        filename = os.path.join(reportconfig['directory'], filename)
+      if os.path.exists(filename):
+        # Running inside docker sometimes doesnt let you write over files. 
+        os.remove(filename)
+      print("Writing out:", filename)
+      fig.write_image(filename)
+      print("Written out:", filename)
```
Comment (apetrynet) [inline enctests/testframework/utils/outputTemplate.py:87]: Should we consider putting this in it's own function? For clarity(?)

**Review by apetrynet — COMMENTED** [review]
It's really nice to see the different parts come together. I've added some generic python comments here and there.

File: enctests/testframework/main.py (hunk 5 above)
Comment (richardssam) [reply enctests/testframework/main.py:407]: I can do the reverse, add a skip-reports flag, it feels like if you have specified the report generation in the file, it should run it by default.

Comment (apetrynet) [reply enctests/testframework/main.py:407]: agree. opt-out rather than opt-in sounds good

File: enctests/testframework/generatetests.py (hunk 1 above)
Comment (richardssam) [reply enctests/testframework/generatetests.py]: Should be addressed.

File: enctests/testframework/generatetests.py [hunk 8]
```diff
@@ -0,0 +1,149 @@
...
+    for root, dirs, files in os.walk(rootdir):
+        for name in files:
+            if name.endswith(".md"):
+                markdownfiles.append(os.path.join(root, name))
+
+    allinfo = []
+
+    for markdownfile in markdownfiles:
+        print("------\n", markdownfile)
+        f = open(markdownfile, "r")
```
[the remark this answers is not in the document]
Comment (richardssam) [reply enctests/testframework/generatetests.py]: Should be addressed.

File: enctests/testframework/generatetests.py (hunk 3 above)
Comment (richardssam) [reply enctests/testframework/generatetests.py]: Should be addressed.

File: enctests/testframework/generatetests.py (hunk 4 above)
Comment (richardssam) [reply enctests/testframework/generatetests.py]: Should be addressed.

File: enctests/testframework/utils/outputTemplate.py (hunk 6 above)
Comment (richardssam) [reply enctests/testframework/utils/outputTemplate.py]: Should be addressed.

File: enctests/testframework/utils/outputTemplate.py (hunk 7 above)
Comment (richardssam) [reply enctests/testframework/utils/outputTemplate.py]: Done.

File: enctests/testframework/generatetests.py [hunk 9]
```diff
@@ -0,0 +1,144 @@
...
+
+    markdownfiles = []
+    for root, dirs, files in os.walk(rootdir):
+        for name in files:
+            if name.endswith(".md"):
+                markdownfiles.append(os.path.join(root, name))
+
+    allinfo = []
+
+    for markdownfile in markdownfiles:
```
Comment (apetrynet) [inline enctests/testframework/generatetests.py:49]: Would it make sense to create a function out of the markdown parsing? And one of the yaml creation below?

File: enctests/testframework/generatetests.py [hunk 10]
```diff
@@ -0,0 +1,144 @@
...
+                    continue
+                if incomment:
+                    info = info + line
+                if incommand:
+                    command = command + line
+
+    # Now we process the scanned data to create the config file.
+    tests = []
+
+    for testcount, test in enumerate(allinfo):
```
Comment (apetrynet) [inline enctests/testframework/generatetests.py:87]: Should we make this a function like the suggested markdown parsing function?

That would leave `main()` quite clean

File: enctests/testframework/main.py [hunk 11]
```diff
@@ -213,6 +222,10 @@ def get_configs(args, root_path, config_type):
 
 
 def tests_only(test_configs):
+    """
+    Scan the test-configs for just the test configurations (since it can be co-mingled with output and source info).
+    Each test mu
```
Comment (apetrynet) [inline enctests/testframework/main.py:227]: I'm not sure I understand what this comment is saying. "Each test mu"

**Review by apetrynet — COMMENTED** [review]
I had one question about a comment, and a suggestion for future readability. Nothing really of such importance it should hold us back. Let me know if you agree or not and we'll land this.

**Pull request merged**